spring-messaging/src/jmh/java/org/springframework/messaging/simp/broker/DefaultSubscriptionRegistryBenchmark.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.messaging.simp.broker;
+
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.simp.SimpMessageHeaderAccessor;
+import org.springframework.messaging.simp.SimpMessageType;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.util.MultiValueMap;
+
+@BenchmarkMode(Mode.Throughput)
+@Fork(value = 1, warmups = 1)
+@Threads(4)
+@Measurement(iterations = 3, time = 2)
+@Warmup(time = 1, iterations = 3)
+public class DefaultSubscriptionRegistryBenchmark {
+
+	@State(Scope.Benchmark)
+	public static class ServerState {
+		@Param("1000")
+		public int numberOfSessions;
+
+		@Param("10")
+		public int subscriptionsPerSession;
+
+		@Param("10")
+		public int numberOfDestinations;
+
+		public DefaultSubscriptionRegistry registry;
+
+		public String[] destinations;
+
+		public String[] sessionIds;
+
+		@Setup(Level.Trial)
+		public void doSetup() {
+			this.registry = new DefaultSubscriptionRegistry();
+
+			this.destinations = IntStream.range(0, this.numberOfDestinations)
+					.mapToObj(i -> "/some/destination/" + i)
+					.toArray(String[]::new);
```

Comment:
There probably should be some pattern destinations.

Reply:
I will add a benchmark for pattern destinations. Actually, there two places where I can put a pattern.
1) new subscription can be a pattern destination
2) other already registered subscriptions can be a pattern
I will add a flag at both places.